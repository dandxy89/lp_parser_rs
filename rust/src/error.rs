//! Comprehensive error types for LP parsing operations.
//!
//! This module provides a structured error hierarchy with detailed context
//! information for debugging and error handling in LP file parsing.

use thiserror::Error;

/// Comprehensive error type for LP parsing operations.
///
/// This error type provides detailed context about parsing failures,
/// including location information and specific error conditions.
#[derive(Error, Debug, Clone, PartialEq, Eq)]
pub enum LpParseError {
    /// Invalid or malformed constraint syntax
    #[error("Invalid constraint syntax at position {position}: {context}")]
    ConstraintSyntax { position: usize, context: String },

    /// Invalid or malformed objective syntax
    #[error("Invalid objective syntax at position {position}: {context}")]
    ObjectiveSyntax { position: usize, context: String },

    /// Unknown or invalid variable type specification
    #[error("Unknown variable type '{var_type}' for variable '{variable}'")]
    UnknownVariableType { variable: String, var_type: String },

    /// Reference to an undefined variable
    #[error("Undefined variable '{variable}' referenced in {context}")]
    UndefinedVariable { variable: String, context: String },

    /// Duplicate definition of a component
    #[error("Duplicate {component_type} '{name}' defined")]
    DuplicateDefinition { component_type: String, name: String },

    /// Invalid numerical value or format
    #[error("Invalid number format '{value}' at position {position}")]
    InvalidNumber { value: String, position: usize },

    /// Missing required section in LP file
    #[error("Missing required section: {section}")]
    MissingSection { section: String },

    /// Invalid bound specification
    #[error("Invalid bounds for variable '{variable}': {details}")]
    InvalidBounds { variable: String, details: String },

    /// Invalid SOS constraint specification
    #[error("Invalid SOS constraint '{constraint}': {details}")]
    InvalidSosConstraint { constraint: String, details: String },

    /// Validation error for logical consistency
    #[error("Validation error: {message}")]
    ValidationError { message: String },

    /// Generic parsing error with context
    #[error("Parse error at position {position}: {message}")]
    ParseError { position: usize, message: String },

    /// File I/O related errors
    #[error("File I/O error: {message}")]
    IoError { message: String },

    /// Internal parser state errors
    #[error("Internal parser error: {message}")]
    InternalError { message: String },
}

impl LpParseError {
    /// Create a new constraint syntax error
    pub fn constraint_syntax(position: usize, context: impl Into<String>) -> Self {
        Self::ConstraintSyntax { position, context: context.into() }
    }

    /// Create a new objective syntax error
    pub fn objective_syntax(position: usize, context: impl Into<String>) -> Self {
        Self::ObjectiveSyntax { position, context: context.into() }
    }

    /// Create a new unknown variable type error
    pub fn unknown_variable_type(variable: impl Into<String>, var_type: impl Into<String>) -> Self {
        Self::UnknownVariableType { variable: variable.into(), var_type: var_type.into() }
    }

    /// Create a new undefined variable error
    pub fn undefined_variable(variable: impl Into<String>, context: impl Into<String>) -> Self {
        Self::UndefinedVariable { variable: variable.into(), context: context.into() }
    }

    /// Create a new duplicate definition error
    pub fn duplicate_definition(component_type: impl Into<String>, name: impl Into<String>) -> Self {
        Self::DuplicateDefinition { component_type: component_type.into(), name: name.into() }
    }

    /// Create a new invalid number error
    pub fn invalid_number(value: impl Into<String>, position: usize) -> Self {
        Self::InvalidNumber { value: value.into(), position }
    }

    /// Create a new missing section error
    pub fn missing_section(section: impl Into<String>) -> Self {
        Self::MissingSection { section: section.into() }
    }

    /// Create a new invalid bounds error
    pub fn invalid_bounds(variable: impl Into<String>, details: impl Into<String>) -> Self {
        Self::InvalidBounds { variable: variable.into(), details: details.into() }
    }

    /// Create a new invalid SOS constraint error
    pub fn invalid_sos_constraint(constraint: impl Into<String>, details: impl Into<String>) -> Self {
        Self::InvalidSosConstraint { constraint: constraint.into(), details: details.into() }
    }

    /// Create a new validation error
    pub fn validation_error(message: impl Into<String>) -> Self {
        Self::ValidationError { message: message.into() }
    }

    /// Create a new parse error
    pub fn parse_error(position: usize, message: impl Into<String>) -> Self {
        Self::ParseError { position, message: message.into() }
    }

    /// Create a new I/O error
    pub fn io_error(message: impl Into<String>) -> Self {
        Self::IoError { message: message.into() }
    }

    /// Create a new internal error
    pub fn internal_error(message: impl Into<String>) -> Self {
        Self::InternalError { message: message.into() }
    }
}

/// Convert from nom parsing errors to our custom error type
impl<'a> From<nom::Err<nom::error::Error<&'a str>>> for LpParseError {
    fn from(err: nom::Err<nom::error::Error<&'a str>>) -> Self {
        match err {
            nom::Err::Incomplete(_) => Self::parse_error(0, "Incomplete input"),
            nom::Err::Error(e) | nom::Err::Failure(e) => Self::parse_error(e.input.as_ptr() as usize, format!("Parse error: {:?}", e.code)),
        }
    }
}

/// Convert from standard I/O errors
impl From<std::io::Error> for LpParseError {
    fn from(err: std::io::Error) -> Self {
        Self::io_error(err.to_string())
    }
}

/// Convert from boxed errors (used by CSV module)
impl From<Box<dyn std::error::Error>> for LpParseError {
    fn from(err: Box<dyn std::error::Error>) -> Self {
        Self::io_error(err.to_string())
    }
}

/// Result type alias for LP parsing operations
pub type LpResult<T> = Result<T, LpParseError>;

/// Context extension trait for adding location information to errors
pub trait ErrorContext<T> {
    /// Add position context to an error
    fn with_position(self, position: usize) -> LpResult<T>;

    /// Add general context to an error
    fn with_context(self, context: &str) -> LpResult<T>;
}

impl<T> ErrorContext<T> for Result<T, LpParseError> {
    fn with_position(self, position: usize) -> Self {
        self.map_err(|mut err| {
            if let LpParseError::ParseError { position: ref mut pos, .. } = &mut err {
                *pos = position;
            }
            err
        })
    }

    fn with_context(self, context: &str) -> Self {
        self.map_err(|err| match err {
            LpParseError::ParseError { position, message } => LpParseError::parse_error(position, format!("{context}: {message}")),
            other => other,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_error_creation() {
        let err = LpParseError::constraint_syntax(42, "missing operator");
        assert_eq!(err.to_string(), "Invalid constraint syntax at position 42: missing operator");
    }

    #[test]
    fn test_error_context() {
        let result: LpResult<()> = Err(LpParseError::parse_error(10, "test error"));
        let with_context = result.with_context("parsing constraint");

        assert!(with_context.is_err());
        assert!(with_context.unwrap_err().to_string().contains("parsing constraint"));
    }

    #[test]
    fn test_io_error_conversion() {
        let io_err = std::io::Error::new(std::io::ErrorKind::NotFound, "file not found");
        let lp_err: LpParseError = io_err.into();

        match lp_err {
            LpParseError::IoError { message } => assert!(message.contains("file not found")),
            _ => panic!("Expected IoError"),
        }
    }
}
