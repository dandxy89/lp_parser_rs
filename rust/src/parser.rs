//! File I/O and content parsing utilities.
//!
//! This module provides functionality for reading LP files from disk
//! and preparing their contents for parsing.
//!

use std::fs::File;
use std::io::{BufReader, Read as _};
use std::path::Path;

use crate::error::{LpParseError, LpResult};

#[inline]
/// Parses the contents of a file at the given path into a string.
///
/// # Arguments
///
/// * `path` - A reference to a `Path` that represents the file path to be parsed.
///
/// # Returns
///
/// A `Result` containing the file contents as a `String` if successful, or an `LpParseError`
/// if the file cannot be opened or read.
///
/// # Errors
///
/// Returns an `LpParseError::IoError` if the file cannot be opened or read.
pub fn parse_file(path: &Path) -> LpResult<String> {
    let file = File::open(path).map_err(|e| LpParseError::io_error(format!("Failed to open file '{}': {}", path.display(), e)))?;

    let mut buf_reader = BufReader::new(file);
    let mut contents = String::new();

    buf_reader
        .read_to_string(&mut contents)
        .map_err(|e| LpParseError::io_error(format!("Failed to read file '{}': {}", path.display(), e)))?;

    Ok(contents)
}
