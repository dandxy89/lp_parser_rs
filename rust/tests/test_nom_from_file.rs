//! This Rust code snippet defines a function to read LP (Linear Programming) files from a resources
//! directory and parse them using the `lp_parser_rs` library.
//!
//! It includes several test functions and a macro to generate tests for various LP files, ensuring they can be parsed correctly.
//!

use std::error::Error;
use std::path::PathBuf;

use lp_parser_rs::parser::parse_file;
use lp_parser_rs::problem::LpProblem;

fn read_file_from_resources(file_name: &str) -> Result<String, Box<dyn Error + 'static>> {
    let mut file_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    file_path.push(format!("resources/{file_name}"));
    let contents = parse_file(&file_path)?;

    Ok(contents)
}

#[test]
fn invalid() {
    let input = read_file_from_resources("invalid.lp").expect("failed to read file from resources");
    assert!(LpProblem::parse(&input).is_err());
}

#[test]
fn nom_fit2d() {
    let input = read_file_from_resources("fit2d.lp").expect("failed to read file from resources");
    let parsed = LpProblem::parse(&input).expect("failed to parse LPProblem");
    assert_eq!(parsed.objective_count(), 1);
    assert_eq!(parsed.constraint_count(), 25);
    assert_eq!(parsed.variable_count(), 10500);

    #[cfg(feature = "serde")]
    insta::assert_yaml_snapshot!(parsed, {
        ".variables" => insta::sorted_redaction(),
        ".objectives" => insta::sorted_redaction(),
        ".constraints" => insta::sorted_redaction(),
    });
}

#[ignore]
#[test]
/// A test used my contributors for assisting with issues
fn nom_debug() {
    let input = read_file_from_resources("infile_comments.lp").expect("failed to read file from resources");
    assert!(LpProblem::parse(&input).is_ok());
}

#[macro_export]
macro_rules! generate_test {
    ($test_name:ident, $file:expr) => {
        #[test]
        #[allow(unused_variables)]
        fn $test_name() {
            let input = read_file_from_resources($file).unwrap();
            let parsed = LpProblem::parse(&input).expect("failed to parse LPProblem");

            #[cfg(feature = "serde")]
            insta::assert_yaml_snapshot!(parsed, {
                ".variables" => insta::sorted_redaction(),
                ".objectives" => insta::sorted_redaction(),
                ".constraints" => insta::sorted_redaction(),
            });
        }
    };
}

generate_test!(pulp, "pulp.lp");
generate_test!(pulp2, "pulp2.lp");
generate_test!(limbo, "limbo.lp");
// generate_test!(semi_continuous, "semi_continuous.lp");
generate_test!(sos, "sos.lp");
generate_test!(test, "test.lp");
generate_test!(test2, "test2.lp");
generate_test!(empty_bounds, "empty_bounds.lp");
generate_test!(blank_lines, "blank_lines.lp");
// generate_test!(optional_labels, "optional_labels.lp");
generate_test!(infile_comments, "infile_comments.lp");
generate_test!(infile_comments2, "infile_comments2.lp");
generate_test!(missing_signs, "missing_signs.lp");
generate_test!(scientific_notation_2, "scientific_notation_2.lp");
generate_test!(output, "output.lp");
generate_test!(output2_1, "output2_1.lp");
generate_test!(output2_2, "output2_2.lp");
generate_test!(output2_3, "output2_3.lp");
generate_test!(output2_4, "output2_4.lp");
generate_test!(complex_names, "complex_names.lp");

// Test files from various open source projects on Github

// Mosek: <https://docs.mosek.com/latest/capi/lp-format.html>
generate_test!(scientific_notation, "scientific_notation.lp");
generate_test!(mosek, "mosek.lp");
generate_test!(mosek_bounds, "mosek_bounds.lp");
generate_test!(lol, "lol.lp");
generate_test!(milo1, "milo1.lp");

// From <https://github.com/asbestian/jplex>
generate_test!(obj3_2cons, "3obj_2cons.lp");

// generate_test!(no_end_section, "no_end_section.lp");
generate_test!(obj_2cons_only_binary_vars, "2obj_2cons_only_binary_vars.lp");
generate_test!(obj_2cons_all_variable_types, "2obj_2cons_all_variable_types.lp");
generate_test!(obj_1cons_all_variables_with_bounds, "1obj_1cons_all_variables_with_bounds.lp");
generate_test!(afiro, "afiro.lp");
generate_test!(afiro_ext, "afiro_ext.lp");
generate_test!(boeing1, "boeing1.lp");
generate_test!(boeing2, "boeing2.lp");
generate_test!(fit1d, "fit1d.lp");
generate_test!(kb2, "kb2.lp");
generate_test!(sc50a, "sc50a.lp");

// From <https://github.com/odow/LPWriter.jl>
generate_test!(model2, "model2.lp");
#[test]
fn corrupt() {
    let input = read_file_from_resources("corrupt.lp").expect("failed to read file from resources");
    assert!(LpProblem::parse(&input).is_err());
}

// From <https://github.com/brymck/lp-parse/tree>
generate_test!(model, "model.lp");

// From <https://github.com/josephcslater/JupyterExamples>
generate_test!(sudoku, "sudoku.lp");

// From <https://github.com/afaf-taik/vehicularFL>
generate_test!(wbm, "wbm.lp");

// From <https://github.com/IBMDecisionOptimization/cplexrunonwml>
generate_test!(diet, "diet.lp");

// From <https://github.com/claudiosa/CCS>
// generate_test!(output_cplex_2, "output_cplex_2.lp");

// From <https://github.com/coin-or/pulp>
generate_test!(american_steel_problem, "AmericanSteelProblem.lp");
generate_test!(beer_distribution_problem, "BeerDistributionProblem.lp");
generate_test!(computer_plant_problem, "ComputerPlantProblem.lp");
generate_test!(whiskas_model_2, "WhiskasModel2.lp");

// From <https://lpsolve.sourceforge.net/5.0/CPLEX-format.htm>
generate_test!(cplex, "cplex.lp");
