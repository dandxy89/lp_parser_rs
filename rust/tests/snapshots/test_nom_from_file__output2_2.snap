---
source: rust/tests/test_nom_from_file.rs
expression: parsed
---
name: "Problem name: ClpDefaultName"
sense: Minimize
objectives:
  obj:
    name: obj
    coefficients:
      - name: x0
        value: 2.1
      - name: x1
        value: 1000
      - name: x2
        value: 2.2
      - name: x3
        value: 1000
      - name: x4
        value: 2.3
      - name: x5
        value: 1000
      - name: x6
        value: 1.1
      - name: x7
        value: 1.2
      - name: x8
        value: 1.3
constraints:
  cons0:
    type: Standard
    name: cons0
    coefficients:
      - name: x0
        value: 1
      - name: x1
        value: 1
    operator: LTE
    rhs: 21.1
  cons1:
    type: Standard
    name: cons1
    coefficients:
      - name: x2
        value: 1
      - name: x3
        value: 1
    operator: LTE
    rhs: 22.2
  cons2:
    type: Standard
    name: cons2
    coefficients:
      - name: x4
        value: 1
      - name: x5
        value: 1
    operator: LTE
    rhs: 23.3
  cons3:
    type: Standard
    name: cons3
    coefficients:
      - name: x6
        value: 1
    operator: LTE
    rhs: 11.1
  cons4:
    type: Standard
    name: cons4
    coefficients:
      - name: x7
        value: 1
    operator: LTE
    rhs: 12.2
  cons5:
    type: Standard
    name: cons5
    coefficients:
      - name: x8
        value: 1
    operator: LTE
    rhs: 13.3
variables:
  x0:
    name: x0
    var_type:
      DoubleBound:
        - 0
        - 21.1
  x1:
    name: x1
    var_type: Free
  x2:
    name: x2
    var_type:
      DoubleBound:
        - 0
        - 22.2
  x3:
    name: x3
    var_type: Free
  x4:
    name: x4
    var_type:
      DoubleBound:
        - 0
        - 23.3
  x5:
    name: x5
    var_type: Free
  x6:
    name: x6
    var_type:
      DoubleBound:
        - 0
        - 11.1
  x7:
    name: x7
    var_type:
      DoubleBound:
        - 0
        - 12.2
  x8:
    name: x8
    var_type:
      DoubleBound:
        - 0
        - 13.3
